File: BotsSDK/korebotsdklib/src/main/java/kore/botssdk/utils/markdown/MarkdownImageTagHandler.java
package kore.botssdk.utils.markdown;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.text.Html;
import android.util.Base64;
import android.widget.TextView;

import java.io.BufferedInputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;



public class MarkdownImageTagHandler implements Html.ImageGetter {

    private final Context context;
    private final TextView htmlTextViewRemote;
    private final String htmlStringRemote;

    public MarkdownImageTagHandler(Context context, TextView htmlTextViewRemote, String htmlStringRemote) {
        this.context = context;
        this.htmlTextViewRemote = htmlTextViewRemote;
        this.htmlStringRemote = htmlStringRemote;
    }

    @Override
    public Drawable getDrawable(String source) {
        HttpGetDrawableTask httpGetDrawableTask = new HttpGetDrawableTask(context, htmlTextViewRemote, htmlStringRemote);
        httpGetDrawableTask.execute(source);
        return null;
    }
}

class HttpGetDrawableTask extends AsyncTask<String, Void, Drawable> {

    private final Context context;
    private final TextView taskTextView;
    private final String taskHtmlString;
    private final int dp1;

    public HttpGetDrawableTask(Context context, TextView v, String s) {
        this.context = context;
        taskTextView = v;
        taskHtmlString = s;
        dp1 = (int)Resources.getSystem().getDisplayMetrics().density;
    }

    @Override
    protected Drawable doInBackground(String... params) {
        Drawable drawable = null;
        URL sourceURL;
        String base64;
        BufferedInputStream bufferedInputStream = null;
        InputStream inputStream = null;
        try
        {
            if(params[0].contains("base64,"))
            {
                base64 = params[0].substring(params[0].indexOf(",") + 1);
                byte[] decodedString = Base64.decode(base64.getBytes(), Base64.DEFAULT);
                Bitmap bm = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);

                // convert Bitmap to Drawable
                drawable = new BitmapDrawable(context.getResources(), bm);

                drawable.setBounds(0, 0, bm.getWidth(), bm.getHeight());
            }
            else
            {
                sourceURL = new URL(params[0]);
                URLConnection urlConnection = sourceURL.openConnection();
                urlConnection.connect();
                inputStream = urlConnection.getInputStream();
                bufferedInputStream = new BufferedInputStream(inputStream);
                Bitmap bm = BitmapFactory.decodeStream(bufferedInputStream);

                // convert Bitmap to Drawable
                drawable = new BitmapDrawable(context.getResources(), bm);
                drawable.setBounds(0, 0, bm.getWidth(), bm.getHeight());
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
        finally {
            try {
                if(inputStream != null) inputStream.close();
                if(bufferedInputStream != null) bufferedInputStream.close();
            }catch (Exception e){e.printStackTrace();}
        }

        return drawable;
    }

    @Override
    protected void onPostExecute(Drawable result) {

        final Drawable taskDrawable = result;

        if (taskDrawable != null) {
            taskTextView.setText(Html.fromHtml(taskHtmlString,
                new Html.ImageGetter() {

                    @Override
                    public Drawable getDrawable(String source) {
                        return taskDrawable;
                    }
                }, null));
        }
    }
}

File: BotSdk/botclient/src/main/java/com/kore/common/markdown/MarkdownImageTagHandler.kt
package com.kore.common.markdown

import android.content.Context
import android.content.res.Resources
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.AsyncTask
import android.text.Html
import android.util.Base64
import android.widget.TextView
import androidx.core.text.HtmlCompat
import java.io.BufferedInputStream
import java.io.InputStream
import java.net.URL

class MarkdownImageTagHandler(
    context: Context,
    htmlTextViewRemote: TextView,
    htmlStringRemote: String
) : Html.ImageGetter {

    val context: Context? = null
    private val htmlTextViewRemote: TextView? = null
    private val htmlStringRemote: String? = null

    override fun getDrawable(source: String?): Drawable? {
        val httpGetDrawableTask = HttpGetDrawableTask(context, htmlTextViewRemote, htmlStringRemote)
        httpGetDrawableTask.execute(source)
        return null
    }
}


internal class HttpGetDrawableTask(
    val context: Context?,
    val taskTextView: TextView?,
    private val taskHtmlString: String?
) :
    AsyncTask<String?, Void?, Drawable?>() {
    private val dp1: Int = Resources.getSystem().displayMetrics.density.toInt()

    override fun doInBackground(vararg params: String?): Drawable? {
        var drawable: Drawable? = null
        val sourceURL: URL
        val base64: String
        var bufferedInputStream: BufferedInputStream? = null
        var inputStream: InputStream? = null
        try {
            if (params[0]?.contains("base64,") == true) {
                base64 = params[0]?.substring(params[0]?.indexOf(",")?.plus(1) ?: 0).toString()
                val decodedString = Base64.decode(base64.toByteArray(), Base64.DEFAULT)
                val bm = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.size)

                // convert Bitmap to Drawable
                drawable = BitmapDrawable(context!!.resources, bm)
                drawable.setBounds(0, 0, bm.width, bm.height)
            } else {
                sourceURL = URL(params[0])
                val urlConnection = sourceURL.openConnection()
                urlConnection.connect()
                inputStream = urlConnection.getInputStream()
                bufferedInputStream = BufferedInputStream(inputStream)
                val bm = BitmapFactory.decodeStream(bufferedInputStream)

                // convert Bitmap to Drawable
                drawable = BitmapDrawable(context!!.resources, bm)
                drawable.setBounds(0, 0, bm.width, bm.height)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                inputStream?.close()
                bufferedInputStream?.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return drawable!!
    }

    override fun onPostExecute(result: Drawable?) {
        if (result != null) {
            taskTextView!!.text = Html.fromHtml(
                taskHtmlString,
                HtmlCompat.FROM_HTML_MODE_LEGACY,
                { result }, null
            )
        }
    }
}
